odoo/addons/base/models/ir_actions_report.py

```diff
@@ -501,25 +501,37 @@ def _get_report_from_name(self, report_name):
         return report_obj.with_context(context).sudo().search(conditions, limit=1)
 
     @api.model
-    def barcode(self, barcode_type, value, width=600, height=100, humanreadable=0, quiet=1, mask=None):
+    def barcode(self, barcode_type, value, width=600, height=100, humanreadable=0, quiet=1, mask=None, **kwargs):
+        bar_border = 4
```

Comment:
```suggestion
    def barcode(self, barcode_type, value, **kwargs):
        defaults = {
            'width': (600, int),
            'height': (100, int),
            'humanreadable': (False, lambda v: bool(int(v))),
            'quiet': (True, lambda v: bool(int(v))),
            'mask': (None, lambda v: v),
            'barLevel': ('L', lambda v: v if v in ('L', 'M', 'Q', 'H') else 'L'),
            'barBorder': (4, int),
        }
        kwargs={k: validator(kwargs.get(k, value)) for k, (value, validator) in defaults.values()}
        kwargs['humanReadable'] = kwargs.pop('humanreadable')
```

Or is this breaking something?

Reply:
> This is v14, so it's stable, some error behaviour would change.

You would be raising a (obfuscated) `ValueError` for the `int` casting anyways. Same error type, different (and more explicit) message.

> We had a long discussion above about how the `width` `height` `humanreadable` and `quiet` are validated inside the `try...except` and will blow up throwing an exception, while an error in `barLevel` should just log a warning. [#81183 (comment)](https://github.com/odoo/odoo/pull/81183#discussion_r767510978)

`ValueError` is `ValueError` why would a wrong value for `width` raise one, while a wrong value for `barLevel` wouldn't?
AFAIK, the casting to `int` is only done because it can come from the non typed arguments from `'/report/barcode'`. If giving a wrong value for `barLevel` should result in silently replacing it with the default, it should be the same for  `width`.


> I just followed the suggestion here: [#81183 (comment)](https://github.com/odoo/odoo/pull/81183#discussion_r767509726) and just _added_ a kwargs parameter.
> I also tried to get some default in a `dict` and was told that it's complexifying the diff

What was done there is not what I'm proposing.

________________

If @xmo-odoo doesn't agree with this, he can r+ :shrug:

Reply:
@william-andre's suggestion is not wrong here: if parsing the width, height, humanreadable, or quiet values fails with a `ValueError`, they're going to fail again when recursively trying to `barcode('Code128', ...)` in the error handler. So them being outside the `try/except` block makes no semantics difference.

Although this:

> `ValueError` is `ValueError`

is not entirely correct, because reportlab's barcode can raise `ValueError`on some barcode configurations.

Reply:
> as indicated has an entire discussion on the subject, the gist of which is "getting the error level of your qrcode wrong usually doesn't matter".

If I am asking for something specific, don't give me something else without telling me. If I specify the `barLevel` but you can't provide it, tell me instead of making something wrong without telling me. Because if I am asking, there must be a reason; which is most likely a (legal) requirement.